pkg/tasker/tasker.go

```diff
@@ -0,0 +1,196 @@
+
+type MiningTasker struct {
+	ctx              context.Context
+	close            context.CancelFunc
+	logger           log.Logger
+	proxy            db.DataServerProxy
+	accounts         []*rpc.Account
+	currChallenge    *mining.MiningChallenge
+	contractInstance *contracts.ITellor
+	client           contracts.ETHClient
+	cfg              *config.Config
+	workSink         chan *mining.Work
+	Running          bool
+	done             chan bool
+	resubscribe      chan bool
+}
+
+func CreateTasker(ctx context.Context, logger log.Logger, cfg *config.Config, proxy db.DataServerProxy, client contracts.ETHClient, contract *contracts.ITellor, accounts []*rpc.Account) (*MiningTasker, chan *mining.Work) {
+	ctx, close := context.WithCancel(ctx)
+	tasker := &MiningTasker{
+		ctx:              ctx,
+		close:            close,
+		proxy:            proxy,
+		accounts:         accounts,
+		contractInstance: contract,
+		workSink:         make(chan *mining.Work, 1),
+		done:             make(chan bool),
+		resubscribe:      make(chan bool),
+		logger:           log.With(logger, "component", ComponentName),
+		cfg:              cfg,
+		client:           client,
+	}
+	return tasker, tasker.workSink
+}
+
+func (mt *MiningTasker) getCurrentChallenge() (*tellor.ITellorNewChallenge, error) {
+	newVariables, err := mt.contractInstance.GetNewCurrentVariables(nil)
+	if err != nil {
+		level.Warn(mt.logger).Log("msg", "new current variables retrieval - contract might not be upgraded", "err", err)
+		return nil, err
+	}
+	return &tellor.ITellorNewChallenge{
+		CurrentChallenge: newVariables.Challenge,
+		Difficulty:       newVariables.Difficutly,
+		CurrentRequestId: newVariables.RequestIds,
+		TotalTips:        newVariables.Tip,
+	}, err
+}
+
+func (mt *MiningTasker) getNewChallengeChannel() (chan *tellor.ITellorNewChallenge, event.Subscription, error) {
+	sink := make(chan *tellor.ITellorNewChallenge)
+	var tellorFilterer *tellor.ITellorFilterer
+	tellorFilterer, err := tellor.NewITellorFilterer(mt.contractInstance.Address, mt.client)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "getting ITellorFilterer instance")
+	}
+	sub, err := tellorFilterer.WatchNewChallenge(&bind.WatchOpts{}, sink, nil)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "getting NewChallenge channel")
+	}
+	return sink, sub, nil
+}
+
+func (mt *MiningTasker) subscribeToNewChallenge() error {
+	sink, sub, err := mt.getNewChallengeChannel()
+	if err != nil {
+		return err
+	}
+	level.Info(mt.logger).Log("msg", "subscribed to NewChallenge events")
+
+	go func() {
+		for {
+			select {
+			case <-mt.done:
+				sub.Unsubscribe()
+				level.Info(mt.logger).Log("msg", "unsubscribed to NewChallenge events")
+				return
+			case err := <-sub.Err():
+				if err != nil {
+					level.Error(mt.logger).Log(
+						"msg",
+						"new challenge subscription error",
+						"err", err)
+				}
+				mt.resubscribe <- true
+			case vLog := <-sink:
+				mt.sendWork(vLog)
+			}
+		}
+	}()
+	return nil
```

Comment:
I think the logic here can be done without any go routines. 
When there is erros just resubscribe and set the new channel., don't see why we need to use goroutines for this. Ping if you can't figure it out and I will have another look.
```suggestion
	go func() {
		for {
			select {
			case <-mt.done:
				sub.Unsubscribe()
				level.Info(mt.logger).Log("msg", "unsubscribed to NewChallenge events")
				return
			case err := <-sub.Err():
				if err != nil {
					level.Error(mt.logger).Log(
						"msg",
						"new challenge subscription error",
						"err", err)
				}
				mt.resubscribe <- true
			case vLog := <-sink:
				mt.sendWork(vLog)
			}
		}
	}()
	return nil
```

Reply:
Ok, I changed the logic. please take a look at that after I committed the changes.